docs/debugging.md

```diff
@@ -0,0 +1,130 @@
+# Debugging Project Verona
+
+We have built several features into the codebase to aid debugging.
+This document explains those features.
+If anything is unclear, please PR fixes or raise issues.
+
+We will split the rest of this document based on the various stages of running a Verona program
+
+* Compiler
+* Interpreter
+* Runtime
+
+## Debugging the compiler
+
+The simplist way to debug the compiler is to get it to dump all the phases of compiling every method:
+```
+veronac.exe --dump-path dmp region101.verona
+```
+This dumps all the stages of compiling each method into a sub-directory `dmp`.
+For each Class and Method, you should find:
+```
+  [Class].[Method].ast.txt
+  [Class].[Method].resolved-ast.txt
+  [Class].[Method].elaborated-ast.txt
+  [Class].[Method].wf-types.txt
+  [Class].[Method].ir.txt
+  [Class].[Method].liveness.txt
+  [Class].[Method].constraints.txt
+  [Class].[Method].infer.txt
+  [Class].[Method].solver.txt
+  [Class].[Method].substitution.txt
+  [Class].[Method].types.txt
+  [Class].[Method].typed-ir.txt
+  [Class].[Method].make-region-graph.txt
+  [Class].[Method].region-graph.txt
+```
+The order is the order of the phases in the compiler, and can be used to track where something has gone wrong.
+
+Individual, phases and results for particular methods can be output using other options to the compiler.
+
+### Debugging the parser
+
+The Verona parser currently uses [Pegmatite](https://github.com/CompilerTeaching/Pegmatite), a PEG parser design the teaching and rapid prototyping.
+This has two compile options that can aid debugging:
+
+* DEBUG_PARSING  - this outputs tracing during parsing the string
+* DEBUG_AST_CONSTRUCTION  - this outputs the construction of the AST nodes
+
+
+## Debugging the interpreter
+
+As the majority of code running in the interpreter is actually the runtime, the subsequent section could also prove useful.
+The interpreter provides a trace of all the instructions that it executes.
+Every option that you can pass to the interpreter, can also be passed to the compiler prepended with `run-`, for example:
+```
+interpreter foo.bc --verbose
+```
+or 
+```
+veronac foo.verona --run --run-verbose
+```
+The `--verbose` option basically provides a trace of the interpreters operations:
+for example:
+```
+veronac.exe --run --run-verbose region101.verona
```

Comment:
Quite a detail, but you use `.exe` in some places but not others. Better to be consistent. I'd say no `.exe`, especially since I think windows doesn't require it.

Reply:
Fixed